### PR TITLE
dubbo_proxy : Add NamedDeserializerConfigFactory definition

### PR DESCRIPTION
**Pull request by gengleilei**
*Description*: Add NamedDeserializerConfigFactory definition
*Risk Level*: low
*Testing*: unit test
*Docs Changes*: N/A
*Release Notes*: N/A
[Optional Fixes #Issue]
[Optional *Deprecated*:]

this is only part of the code for DubboProxy, the rest of the code hasn't been committed yet, complete implementation: https://github.com/gengleilei/envoy/tree/feature-dubbo-router-develop/source/extensions/filters/network/dubbo_proxy

### CODE REVIEW

**Comment by lizan** [comment]
@zyfjeff can you take first pass?

**Comment by zyfjeff** [comment]
@lizan Sorry, this submission should take the initiative to take the first pass. 

@gengleilei  At the back of the submitted "/review @zyfjeff", please

File: source/extensions/filters/network/dubbo_proxy/deserializer.h [hunk 1]
```diff
@@ -1,23 +1,52 @@
...
 namespace Extensions {
 namespace NetworkFilters {
 namespace DubboProxy {
 
+enum class DeserializerType : uint8_t {
+  Hessian = 0,
+  Json = 1,
+
+  // ATTENTION: MAKE SURE THIS REMAINS EQUAL TO THE LAST PROTOCOL TYPE
+  LastDeserializerType = Json,
```
Comment (zyfjeff) [inline source/extensions/filters/network/dubbo_proxy/deserializer.h]: Didn't see where it was used, can it be removed here?

File: source/extensions/filters/network/dubbo_proxy/deserializer.h [hunk 2]
```diff
@@ -1,23 +1,52 @@
...
+#include "common/common/assert.h"
+#include "common/config/utility.h"
 #include "common/singleton/const_singleton.h"
 
 namespace Envoy {
 namespace Extensions {
 namespace NetworkFilters {
 namespace DubboProxy {
 
+enum class DeserializerType : uint8_t {
```
Comment (zyfjeff) [inline source/extensions/filters/network/dubbo_proxy/deserializer.h]: I think there should be a reuse `SerializationType`

https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/network/dubbo_proxy/protocol.h#L53

File: source/extensions/filters/network/dubbo_proxy/deserializer.h [hunk 3]
```diff
@@ -1,23 +1,52 @@
...
 public:
-  // hessian deserializer
-  const std::string Hessian = "hessian";
-  // json deserializer
-  const std::string Json = "json";
+  struct DeserializerTypeHash {
+    template <typename T> std::size_t operator()(T t) const { return static_cast<std::size_t>(t); }
+  };
+
+  typedef std::unordered_map<DeserializerType, std::string, DeserializerTypeHash>
```
Comment (zyfjeff) [inline source/extensions/filters/network/dubbo_proxy/deserializer.h]: ditto, Ideally should be through the serialization type to find the corresponding deserializer

File: source/extensions/filters/network/dubbo_proxy/deserializer.h [hunk 4]
```diff
@@ -82,6 +117,55 @@ class Deserializer {
...
+   * produced by the factory.
+   */
+  virtual std::string name() PURE;
+
+  /**
+   * Convenience method to lookup a factory by type.
+   * @param TransportType the transport type
+   * @return NamedDeserializerConfigFactory& for the TransportType
+   */
+  static NamedDeserializerConfigFactory& getFactory(DeserializerType type) {
```
Comment (zyfjeff) [inline source/extensions/filters/network/dubbo_proxy/deserializer.h]: ditto

File: source/extensions/filters/network/dubbo_proxy/deserializer.h (hunk 1 above)
Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/deserializer.h]: ok, i deleted it

File: source/extensions/filters/network/dubbo_proxy/deserializer.h (hunk 2 above)
Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/deserializer.h]: I have modified it.

File: source/extensions/filters/network/dubbo_proxy/deserializer.h (hunk 3 above)
Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/deserializer.h]: I have modified it.

File: source/extensions/filters/network/dubbo_proxy/deserializer.h (hunk 4 above)
Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/deserializer.h]: modified

**Comment by zyfjeff** [comment]
LGTM @lizan any comments?

File: source/extensions/filters/network/dubbo_proxy/deserializer.h [hunk 5]
```diff
@@ -82,6 +111,56 @@ class Deserializer {
...
+  }
+};
+
+/**
+ * DeserializerFactoryBase provides a template for a trivial NamedDeserializerConfigFactory.
+ */
+template <class DeserializerImpl>
+class DeserializerFactoryBase : public NamedDeserializerConfigFactory {
+  DeserializerPtr createDeserializer() override {
+    return std::move(std::make_unique<DeserializerImpl>());
```
Comment (lizan) [inline source/extensions/filters/network/dubbo_proxy/deserializer.h]: no need of `std::move`.

Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/deserializer.h]: I deleted it

File: DEPRECATED.md [hunk 6]
```diff
@@ -6,7 +6,9 @@ As of release 1.3.0, Envoy will follow a
 The following features have been DEPRECATED and will be removed in the specified release cycle.
 A logged warning is expected for each deprecated item that is in deprecation window.
 
-## Version 1.9.0 (pending)
+## Version 1.10.0 (pending)
```
Comment (lizan) [inline DEPRECATED.md]: not sure why those appears on diff, they shouldn't, can you merge master again?

Comment (gengleilei) [reply DEPRECATED.md]: done

**Comment by zyfjeff** [comment]
@gengleilei Your this PR looks wrong, your PR should not contain the latest content in the master.

**Pull request merged**